akka-persistence-typed/src/main/scala/akka/persistence/typed/EventSourcedSignal.scala

```diff
@@ -52,13 +53,45 @@ final case class SnapshotFailed(metadata: SnapshotMetadata, failure: Throwable)
   def getSnapshotMetadata(): SnapshotMetadata = metadata
 }
 
+object SnapshotMetadata {
+
+  /**
+   * @param persistenceId id of persistent actor from which the snapshot was taken.
+   * @param sequenceNr sequence number at which the snapshot was taken.
+   * @param timestamp time at which the snapshot was saved, defaults to 0 when unknown.
+   *                  in milliseconds from the epoch of 1970-01-01T00:00:00Z.
+   */
+  def apply(persistenceId: String, sequenceNr: Long, timestamp: Long): SnapshotMetadata =
+    new SnapshotMetadata(persistenceId, sequenceNr, timestamp)
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def fromUntyped(metadata: akka.persistence.SnapshotMetadata): SnapshotMetadata =
+    new SnapshotMetadata(metadata.persistenceId, metadata.sequenceNr, metadata.timestamp)
+}
+
+/**
+ * Snapshot metadata.
+ *
+ * @param persistenceId id of persistent actor from which the snapshot was taken.
+ * @param sequenceNr sequence number at which the snapshot was taken.
+ * @param timestamp time at which the snapshot was saved, defaults to 0 when unknown.
+ *                  in milliseconds from the epoch of 1970-01-01T00:00:00Z.
+ */
+final class SnapshotMetadata(val persistenceId: String, val sequenceNr: Long, val timestamp: Long) {
+  override def toString: String =
+    s"SnapshotMetadata($persistenceId,$sequenceNr,$timestamp)"
+}
```

Comment:
Just curious why you chose `class` vs `final case class`

Reply:
Strange, I'm pretty sure I answered yesterday but that seems gone now. Once again, for the records, since it's important.

Case classes can't be changed (e.g. adding a field) in a binary compatible way and therefore we try to stay away from them in public APIs apart from small trivial things that are unlikely to change.